src/mlang/backend_ir/bir.mli

```diff
@@ -61,6 +61,8 @@ type program = {
   mpp_functions : mpp_function FunctionMap.t;
   rules_and_verifs : rule_or_verif ROVMap.t;
   main_function : function_name;
+  context_function : mpp_function;
+  context_with_reset_function : mpp_function;
```

Comment:
I'd rather have those out of the function map, since it corresponds to nothing in the mpp. Maybe a option would be to put the statements lists out in the record (or a context subrecord) with names such as e.g. `undefined_inputs_init` `constant_inputs_init` and `adhoc_spec_tests`. Then make sure the proper statements list are build by function below and that the rest of the code refer to those functions instead of tapping into the program record.

Reply:
I wouldn't worry for OIR as long as you input the same date as before (and even then...)

Reply:
The point with OIR is: it transforms Bir.program to Oir.program and back, this could be solved easily by adding relevant information to Oir.program. 
But I don't know yet if the context part of the statements were included in the optimisation. I think so, because it was added to the function map and indistinguishable from the original statements of the program. If so, my work will have changed the optimisation behaviour.

Reply:
Optimisations already lose information by burying it in its output code. If you give the same input (with context IIRC), then you'd have the same output, losing context information in the process, sure, but its output should not change.

Reply:
If I keep this behaviour, the OIR will effectively bury the potential context in the mpp_functions map of the Bir.program. I feel like this would be an unwanted behaviour as part of the point of having the context isolated in its own field is to keep the ability to choose in the backends if we want to use a contextualized main function or not.

So my position would be that OIR should keep either keep intact the context statements (but then they won't be optimized), or the choice to have them should be made before OIR call, as far as I can tell it's not trivial with the current Driver.

I'm inclined to modify OIR.program to pass context information transparently as the context is not the most useful in the cases we want to optimise the generated code (like C for batch processing).

Reply:
Let it drown, without the context some passes might have the wrong behaviour, and you don't want to put your nose in there for this PR. 

I agree this is unsatisfactory but it's not worth spending much time on this for now. We'll discuss this in more details later.

Reply:
> Let it drown,

Well, now we can't, because optimisation made it back to master and so this PR breaks it as is, so I'm trying to resolve it while staying close of the old behaviour.